File: commons-text/src/test/java/test/pl/wrzasq/commons/text/html/UtilsTest.java
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2015 - 2016, 2018 - 2019 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.wrzasq.commons.text.html;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import pl.wrzasq.commons.text.Formatter;
import pl.wrzasq.commons.text.TextProcessingException;
import pl.wrzasq.commons.text.formatter.FormatterInterface;
import pl.wrzasq.commons.text.html.Utils;

@ExtendWith(MockitoExtension.class)
public class UtilsTest {
    @Mock
    private FormatterInterface formatHandler;

    @Test
    public void firstParagraph() {
        // just for code coverage
        new Utils();

        var value = "foo <span>bar</span>\n<em>baz</em>";
        var html = "<p>" + value + "</p> <p>\nquux</p>";

        Assertions.assertEquals(
            value,
            Utils.firstParagraph(html),
            "Utils.firstParagraph() should return content of the first paragraph from given snippet."
        );
    }

    @Test
    public void firstParagraphEmpty() {
        var value = "foo <span>bar</span>\n<em>baz</em>";

        Assertions.assertEquals(
            "",
            Utils.firstParagraph(value),
            "Utils.firstParagraph() should return empty string if there is no paragraph in given snippet."
        );
    }

    @Test
    public void truncateShortText() {
        var text = "Hello";

        Assertions.assertEquals(
            text,
            Utils.truncate(text, 7),
            "Utils.truncate() should leave text untouched if it't length fits given requirement."
        );
    }

    @Test
    public void truncate() {
        var text = "Hello world!";

        Assertions.assertEquals(
            "Hello…",
            Utils.truncate(text, 7),
            "Utils.truncate() should truncate the text and apply default suffix indicating more content with respect to word bounds."
        );
    }

    @Test
    public void truncateWordBounds() {
        var text = "Hello world!";

        Assertions.assertEquals(
            "Hello w…",
            Utils.truncate(text, 7, false),
            "Utils.truncate() should ignore word bounds if specified to."
        );
    }

    @Test
    public void truncateWordBoundsNoWord() {
        var text = "Helloworld!";

        Assertions.assertEquals(
            "Hellowo…",
            Utils.truncate(text, 7, true),
            "Utils.truncate() should ignore word bounds if a word bound can't be found."
        );
    }

    @Test
    public void truncateSuffix() {
        var text = "Hello world!";

        Assertions.assertEquals(
            "Hello.",
            Utils.truncate(text, 7, "."),
            "Utils.truncate() should use given suffix if specified."
        );
    }

    @Test
    public void truncateSuffixWordBounds() {
        var text = "Hello world!";

        Assertions.assertEquals(
            "Hello w.",
            Utils.truncate(text, 7, ".", false),
            "Utils.truncate() should use given suffix and ignore word bounds if specified to."
        );
    }

    @Test
    public void format()
        throws TextProcessingException {
        var formatter = new Formatter();
        formatter.registerFormatter("foo", this.formatHandler);

        var input = "bar";
        var result = "baz";

        Mockito.when(this.formatHandler.transform(input)).thenReturn(result);

        Utils.setFormatter(formatter);

        Assertions.assertEquals(
            result,
            Utils.format("foo", input),
            "Utils.format() should return text formatted by given format handler."
        );

        Mockito.verify(this.formatHandler).transform(input);
    }

    @Test
    public void urlEncode() {
        Assertions.assertEquals(
            "foo+bar%2F%E2%80%A6",
            Utils.urlEncode("foo bar/…"),
            "Utils.urlEncode() should build URL-suitable string from the given input."
        );
    }
}

File: commons-text/src/test/kotlin/test/pl/wrzasq/commons/text/html/UtilsTest.kt
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2015 - 2016, 2018 - 2019, 2021 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.wrzasq.commons.text.html

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import pl.wrzasq.commons.text.html.Utils.firstParagraph
import pl.wrzasq.commons.text.html.Utils.truncate
import pl.wrzasq.commons.text.html.Utils.setFormatter
import pl.wrzasq.commons.text.html.Utils.format
import pl.wrzasq.commons.text.html.Utils.urlEncode
import org.junit.jupiter.api.extension.ExtendWith
import pl.wrzasq.commons.text.formatter.FormatterInterface
import kotlin.Throws
import pl.wrzasq.commons.text.TextProcessingException
import pl.wrzasq.commons.text.Formatter

private const val HELLO_WORLD = "Hello world!"

@ExtendWith(MockKExtension::class)
class UtilsTest {
    @MockK
    lateinit var formatHandler: FormatterInterface

    @Test
    fun firstParagraph() {
        val value = "foo <span>bar</span>\n<em>baz</em>"
        val html = "<p>$value</p> <p>\nquux</p>"
        assertEquals(value, firstParagraph(html))
    }

    @Test
    fun firstParagraphEmpty() {
        val value = "foo <span>bar</span>\n<em>baz</em>"
        assertEquals("", firstParagraph(value))
    }

    @Test
    fun truncateShortText() {
        val text = "Hello"
        assertEquals(text, truncate(text, 7))
    }

    @Test
    fun truncate() {
        assertEquals("Hello…", truncate(HELLO_WORLD, 7))
    }

    @Test
    fun truncateWordBounds() {
        assertEquals("Hello w…", truncate(HELLO_WORLD, 7, false))
    }

    @Test
    fun truncateWordBoundsNoWord() {
        val text = "Helloworld!"
        assertEquals("Hellowo…", truncate(text, 7, true))
    }

    @Test
    fun truncateSuffix() {
        assertEquals("Hello.", truncate(HELLO_WORLD, 7, "."))
    }

    @Test
    fun truncateSuffixWordBounds() {
        assertEquals("Hello w.", truncate(HELLO_WORLD, 7, ".", false))
    }

    @Test
    @Throws(TextProcessingException::class)
    fun format() {
        val formatter = Formatter()
        formatter.registerFormatter("foo", formatHandler)
        val input = "bar"
        val result = "baz"

        every {
            formatHandler.transform(input)
        } returns result

        setFormatter(formatter)
        assertEquals(result, format("foo", input))

        verify {
            formatHandler.transform(input)
        }
    }

    @Test
    fun urlEncode() {
        assertEquals("foo+bar%2F%E2%80%A6", urlEncode("foo bar/…"))
    }
}
